### PR TITLE
Being refactored: ECN additions

### PR DESCRIPTION
**Pull request by larseggert**
This PR is getting too large. I'm in the process of refactoring it into smaller chunks:

- [x] #1568
- [x] #1604 
- [ ] #1678

### CODE REVIEW

File: Cargo.toml
Comment (martinthomson) [inline Cargo.toml:18]: We won't be able to land a change that pulls from a private branch, at least not without an extraordinary approval process.  I assume that you are in the process of landing changes upstream.

File: neqo-helper/src/lib.rs
Comment (martinthomson) [inline neqo-helper/src/lib.rs:30]: The alternative to creating a new crate is to add this to a new module in neqo-common.  A non-default feature flag that makes this module conditionally available would ensure that mainline code doesn't need to pull in the nix and libc dependencies.

A new crate isn't that big of a deal if you only use it for the command-line tools, but I can see how we might want to use these facilities in Firefox proper, for which having it in neqo-common might help.

File: neqo-helper/src/lib.rs
Comment (martinthomson) [inline neqo-helper/src/lib.rs:100]:
```suggestion
        _ => unreachable!(),
```

I think that we're safe with that in this case.  IPv10 is still a ways off.

File: neqo-helper/src/lib.rs
Comment (martinthomson) [inline neqo-helper/src/lib.rs:103]: You should be able to do just one match.  You are repeating the `addr.as_sockaddr_in[6]().unwrap()` part as it is.

File: neqo-common/src/datagram.rs
Comment (martinthomson) [inline neqo-common/src/datagram.rs:25]:
```suggestion
#[derive(Copy, Clone, PartialEq, Eq)]
```

File: neqo-common/src/datagram.rs
Comment (martinthomson) [inline neqo-common/src/datagram.rs:15]:
```suggestion
#[derive(Copy, Clone, PartialEq, Eq)]
```

File: neqo-helper/src/lib.rs
Comment (martinthomson) [inline neqo-helper/src/lib.rs:117]: How often do we have a raw fd when calling this function?  Can this function call `to_raw_fd()` itself?

File: neqo-helper/src/lib.rs
Comment (martinthomson) [inline neqo-helper/src/lib.rs:68]: See below regarding `fd`

**Review by martinthomson — COMMENTED** [review]
I take it that this is groundwork for ECN

File: Cargo.toml
Comment (larseggert) [reply Cargo.toml:18]: Yep. This is just temporary so the CI runs. I will remove this once upstreams are updated.

File: neqo-helper/src/lib.rs
Comment (larseggert) [reply neqo-helper/src/lib.rs:117]: I couldn't easily figure out how, because it's being called with a `mio::net::UdpSocket` from the server and a `std::net::UdpSocket` from the client. Will investigate.

File: neqo-helper/src/lib.rs
Comment (larseggert) [reply neqo-helper/src/lib.rs:30]: Thanks for the feedback. Will need to read up on those concepts.

File: neqo-helper/src/lib.rs
Comment (martinthomson) [reply neqo-helper/src/lib.rs:117]: Oh, I missed that.  Unless the two implement a shared trait, you won't be able to.  But then it is possible to have that happen.  Whether it is worthwhile, I'll leave as an exercise :)

The basic idea here would be that you have two types that can feed in, so you would have a function signature that is generic:

```rust
pub fn recv_datagram<S: SocketTrait>(socket: S, ...) {
}
```

Then, to allow you to pass in either form, you would define that new trait with the functions you need as follows:

```rust
pub trait SocketTrait { // don't use my name
  fn raw_fd(&self) -> i32;
}
```

You can then implement that trait for the different types you want to support, like so:

```rust
impl SocketTrait for std::net::UdpSocket {
  fn raw_fd(&self) -> i32 { self.get_raw_fd() }
}


impl SocketTrait for mio::net::UdpSocket {
  fn raw_fd(&self) -> i32 { self.get_raw_fd() }
}
```

This means that your code can have multiple instantiations, one for each concrete type that implements this new trait.  In practice, you want one instantiation in the client and another in the server, so there is no code bloat.

This would make the interface more ergonomic, but it's not clear that we really need that capability in this instance.  I made the comment without realizing that there were different types involved.

File: neqo-helper/src/lib.rs
Comment (larseggert) [reply neqo-helper/src/lib.rs]: @martinthomson I think I made that change.

File: neqo-common/src/socket.rs
Comment (martinthomson) [inline neqo-common/src/socket.rs:182]: I would recommend extracting these two blocks into (private) functions.

File: neqo-server/src/main.rs
Comment (martinthomson) [inline neqo-server/src/main.rs:715]: Do you want to open an issue to track EAGAIN handling in these?

File: neqo-common/src/socket.rs
Comment (martinthomson) [inline neqo-common/src/socket.rs:10]: `AsRawFd` = win.

File: neqo-common/src/socket.rs
Comment (martinthomson) [inline neqo-common/src/socket.rs:121]: These uses of `unwrap()` still concern me a little; convince me that this is rather than wrapping the error in a new type.  You aren't using the `io::Result::Err` branch on this either.

File: neqo-common/src/socket.rs
Comment (martinthomson) [inline neqo-common/src/socket.rs:62]:
```suggestion
            debug_assert!(res.is_ok());
```

If this fails, let's not worry too much.

File: neqo-common/src/socket.rs
Comment (martinthomson) [inline neqo-common/src/socket.rs:69]: As above, ...

**Review by martinthomson — DISMISSED** [review]
I've a question for you and some suggestions, but I trust you to resolve those from here (and I don't want to block you while I'm on leave).

File: neqo-common/src/socket.rs
Comment (larseggert) [reply neqo-common/src/socket.rs:49]: This is probably because I am new to Rust, but I don't see _how_ this can return an error? AFAICT it can only panic?

File: neqo-common/src/socket.rs
Comment (martinthomson) [inline neqo-common/src/socket.rs:150]: ???

Comment (larseggert) [reply neqo-common/src/socket.rs:150]: Leftover debug thing; am still trying to get this to build/run on Windows.

**Comment by larseggert** [comment]
@martinthomson could you re-review? Made a bunch of changes that make things work on all platforms again.

**Comment by larseggert** [comment]
Blocked until https://github.com/nix-rust/nix/pull/2203 has shipped.

File: neqo-transport/src/recovery.rs
Comment (martinthomson) [inline neqo-transport/src/recovery.rs:415]: This sort of change really needs to be in a separate patch so that we can look at the recovery tests in more detail.

File: neqo-common/Cargo.toml
Comment (larseggert) [reply neqo-common/Cargo.toml:25]: It's in progress. Depends on another PR upstream.

File: neqo-common/src/datagram.rs
Comment (larseggert) [reply neqo-common/src/datagram.rs]: I think I did this elsewhere, must have missed this one.

File: neqo-common/src/socket.rs
Comment (larseggert) [reply neqo-common/src/socket.rs:399]: ACK

File: neqo-transport/src/recovery.rs
Comment (larseggert) [reply neqo-transport/src/recovery.rs:415]: Yeah, I realized this is getting big. I will break it into individual chunks.

**Comment by mxinden** [comment]
Crossposting https://github.com/mxinden/neqo/pull/1#pullrequestreview-1892959112 here to ensure review on fork does not get lost.

**Comment by larseggert** [comment]
Nothing left in this PR that would be of interest.

**Pull request closed without merging**
